### PR TITLE
Upgrade BCFIPS to 2.0

### LINKED ISSUE #30415: Upgrade BCFIPS to 2.0 to make sure it can be used on OpenJDK 21
According to [1], the 2.0 stream is working with Java 21, when previous 1.0 stream just with Java 17. Also there is recent FIPS 140-3 certification [2], which applies to 2.0 stream (maybe for 1.0 stream as well, but this is not very clear to me).

I don't know exactly, but I estimate that upgrade won't be so huge effort considering that all our BCFIPS integration is in the module "crypto/fips1402" in the codebase. Moreover, we don't provide support for BCFIPS bits and hence we're not dependent on using same versions as Quarkus uses.

[1] https://www.bouncycastle.org/download/bouncy-castle-java-fips/
[2] https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4743

### Original description

It seems like Keycloak has released its v25 and now supports JDK21 and will deprecate JDK 17. That brings me to the question that as of now it seems BouncyCastle FIPS support up to JDK17 so it can affect the Keycloak-FIPS compliance here https://www.keycloak.org/server/fips

What are our plans for that and should we wait for bumping to JDK21 till we have BCFIPS available for it? 
https://github.com/bcgit/bc-java/issues/1558#issuecomment-1882382781

### PR DESCRIPTION
**Pull request by mposolda**
closes #30415

### CODE REVIEW

**Review by jonkoops — DISMISSED** [review]
LGTM

**Review by jonkoops — DISMISSED** [review]

**Review by rmartinc — APPROVED** [review]
So it was integrated with no major issues. Great!

**Review by ahus1 — DISMISSED** [review]
Approving based on @rmartinc's review.

**Comment by ahus1** [comment]
@mposolda - that's great news! Does this mean we can drop JDK17 for KC26 server if this is merged before the release?

**Comment by jonkoops** [comment]
> that's great news! Does this mean we can drop JDK17 for KC26 server if this is merged before the release?

Considering the LTS support we need to provide for Keycloak 26 that would be really nice.

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc [hunk 1]
```diff
@@ -300,6 +300,11 @@ The new event types are supported by the Email Event Listener.
 
 The following event types are now deprecated and will be removed in a future version: `UPDATE_PASSWORD`, `UPDATE_PASSWORD_ERROR`, `UPDATE_TOTP`, `UPDATE_TOTP_ERROR`, `REMOVE_TOTP`, `REMOVE_TOTP_ERROR`
 
+= BouncyCastle FIPS updated
+
+Our FIPS 140-2 integration is now tested and supported with version 2 of BouncyCastle FIPS libraries. This version is certified with Java 21. If you use FIPS 140-2 integration, it is recommended to
```
Comment (rmartinc) [inline docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:305]: Just a little comment, with the upgrade to 2.0 the bcfips is also certified in fips 140-3 (see https://www.bouncycastle.org/resources/bouncy-castle-crypto-api-attains-fips-140-3-certification/). Maybe we can change this paragraph.

Comment (mposolda) [reply docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:305]: @rmartinc This is good point. The thing is, that we support also RHEL8, which itself is only FIPS 140-2, but RHEL9 is FIPS 140-3. So I've instead added additional paragraph about FIPS 140-3 when on RHEL9 system. What do you think?

**Comment by mposolda** [comment]
> @mposolda - that's great news! Does this mean we can drop JDK17 for KC26 server if this is merged before the release?

@ahus1 Good point! There is https://github.com/keycloak/keycloak/issues/31101 for drop OpenJDK 17. Apart from this one, there is policy-enforcer task, but that one is client library and hence unrelated to the server. So maybe we can unless there is something else? (Started thread on the team ML).

**Review by andymunro — APPROVED** [review]
Looks good, @mposolda

**Comment by mposolda** [comment]
@rmartinc @jonkoops @andymunro @ahus1 Thanks everyone!

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc (hunk 1 above)
Comment (rmartinc) [reply docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:305]: Yep, I didn't think about the JVM which should also be fips 140-3. It's OK to me.

**Pull request merged**